plutus-ir/src/Language/PlutusIR/Compiler/Recursion.hs

```diff
@@ -85,19 +88,42 @@ mkFixpoint bs = do
             Nothing  -> lift $ throwing _Error $ CompilationError (PLC.typeAnn ty) "Recursive values must be of function type"
 
     -- See Note [Extra definitions while compiling let-bindings]
-    let arity = fromIntegral $ length funs
-        key = FixpointCombinator arity
+    let
+        fixByKey = FixBy
+        arity = fromIntegral $ length funs
+        fixNKey = FixpointCombinator arity
+
+    -- TODO: reduce duplication
+    fixBy <- do
+        maybeFix <- lookupTerm p0 fixByKey
+        case maybeFix of
+            Just f -> pure f
+            Nothing -> do
+                name <- liftQuote $ toProgramName fixByKey
+                let (fixByTerm, fixByType) = Function.fixByAndType
+                    var :: PLC.VarDecl TyName Name (Provenance a)
+                    var = PLC.VarDecl NoProvenance (NoProvenance <$ name) (NoProvenance <$ fixByType)
+                defineTerm fixByKey (PLC.Def var (NoProvenance <$ fixByTerm, Strict)) mempty
+
+                pure $ PIR.mkVar p0 var
+
     fixN <- do
-        maybeFix <- lookupTerm p0 key
+        maybeFix <- lookupTerm p0 fixNKey
         case maybeFix of
             Just f -> pure f
             Nothing -> do
-                name <- liftQuote $ toProgramName key
-                let (fixNTerm, fixNType) = Function.fixN arity
+                name <- liftQuote $ toProgramName fixNKey
+                let (fixNTerm, fixNType) =
+                        if arity == 1 then Function.fixAndType else Function.fixNAndType arity (void fixBy)
                     var :: PLC.VarDecl TyName Name (Provenance a)
                     var = PLC.VarDecl NoProvenance (NoProvenance <$ name) (NoProvenance <$ fixNType)
-                defineTerm key (PLC.Def var (NoProvenance <$ fixNTerm, Strict)) mempty
+                -- fixN depends on fixBy
+                defineTerm fixNKey (PLC.Def var (NoProvenance <$ fixNTerm, Strict)) (Set.singleton fixByKey)
 
                 pure $ PIR.mkVar p0 var
 
-    liftQuote $ Function.getMutualFixOf p0 fixN funs
+    liftQuote $ case funs of
+        -- Takes a list of function defs and function bodies and turns them into a Scott-encoded tuple, which
+        -- happens to be exactly what we want
+        f :| [] -> Tuple.getSpineToTuple p0 [(PLC.functionDefToType f, Function.getSingleFixOf p0 fixN f)]
+        f :| fs -> Function.getMutualFixOf p0 fixN (f:fs)
```

Comment:
I'd put this logic on the PLC side, but you already know that :)

Reply:
I think this interface point has to be here while the sharing is happening on this side, since you need to thread through the appropriate fixBy/fixNs.

Reply:
It could be `fixN = \fix1 fixNplus1 -> ...` on the PLC side, but yeah, doesn't worth it.